File: Game-1-modular/entities/status_effect.py
"""
Status Effect System
Handles DoT, CC, buffs/debuffs, and other status effects from the tag system
"""

from dataclasses import dataclass, field
from typing import Dict, Any, Optional
from abc import ABC, abstractmethod


@dataclass
class StatusEffect(ABC):
    """
    Base class for all status effects

    Status effects are temporary conditions applied to entities that modify
    their behavior, stats, or apply damage/healing over time.
    """
    status_id: str                      # Unique ID for this status type (e.g., "burn")
    name: str                           # Display name
    duration: float                     # Total duration in seconds
    duration_remaining: float           # Time remaining in seconds
    stacks: int = 1                     # Current stack count
    max_stacks: int = 1                 # Maximum allowed stacks
    source: Optional[Any] = None        # Source entity that applied this
    params: Dict[str, Any] = field(default_factory=dict)  # Effect-specific parameters

    def update(self, dt: float, target: Any) -> bool:
        """
        Update status effect timer and apply periodic effects

        Args:
            dt: Delta time in seconds
            target: Entity this effect is applied to

        Returns:
            True if effect is still active, False if expired
        """
        self.duration_remaining -= dt

        # Apply periodic effect (DoT, HoT, etc.)
        self._apply_periodic_effect(dt, target)

        return self.duration_remaining > 0

    @abstractmethod
    def on_apply(self, target: Any):
        """Called when effect is first applied to target"""
        pass

    @abstractmethod
    def on_remove(self, target: Any):
        """Called when effect is removed from target"""
        pass

    @abstractmethod
    def _apply_periodic_effect(self, dt: float, target: Any):
        """Apply periodic effects like damage or healing"""
        pass

    def add_stack(self, amount: int = 1):
        """Add stacks to this effect (up to max_stacks)"""
        self.stacks = min(self.stacks + amount, self.max_stacks)

    def refresh_duration(self, new_duration: Optional[float] = None):
        """Refresh duration to original or specified value"""
        if new_duration is not None:
            self.duration = new_duration
            self.duration_remaining = new_duration
        else:
            self.duration_remaining = self.duration

    def get_progress_percent(self) -> float:
        """Get percentage of duration remaining (for UI)"""
        if self.duration <= 0:
            return 0.0
        return max(0.0, min(1.0, self.duration_remaining / self.duration))


# ============================================================================
# DAMAGE OVER TIME (DoT) EFFECTS
# ============================================================================

class BurnEffect(StatusEffect):
    """Fire damage over time"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="burn",
            name="Burning",
            duration=duration,
            duration_remaining=duration,
            max_stacks=params.get('burn_max_stacks', 3),
            source=source,
            params=params
        )
        self.damage_per_second = params.get('burn_damage_per_second', 5.0)

    def on_apply(self, target: Any):
        """Visual: Set entity on fire"""
        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('burn')

    def on_remove(self, target: Any):
        """Visual: Remove fire"""
        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('burn')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """Apply fire damage"""
        damage = self.damage_per_second * self.stacks * dt
        if hasattr(target, 'take_damage'):
            target.take_damage(damage, 'fire')
        elif hasattr(target, 'current_health'):
            target.current_health = max(0, target.current_health - damage)


class BleedEffect(StatusEffect):
    """Physical damage over time"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="bleed",
            name="Bleeding",
            duration=duration,
            duration_remaining=duration,
            max_stacks=params.get('bleed_max_stacks', 5),
            source=source,
            params=params
        )
        self.damage_per_second = params.get('bleed_damage_per_second', 3.0)

    def on_apply(self, target: Any):
        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('bleed')

    def on_remove(self, target: Any):
        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('bleed')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """Apply bleed damage"""
        damage = self.damage_per_second * self.stacks * dt
        if hasattr(target, 'take_damage'):
            target.take_damage(damage, 'physical')
        elif hasattr(target, 'current_health'):
            target.current_health = max(0, target.current_health - damage)


class PoisonEffect(StatusEffect):
    """Poison damage over time"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="poison",
            name="Poisoned",
            duration=duration,
            duration_remaining=duration,
            max_stacks=params.get('poison_max_stacks', 10),
            source=source,
            params=params
        )
        self.damage_per_second = params.get('poison_damage_per_second', 2.0)

    def on_apply(self, target: Any):
        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('poison')

    def on_remove(self, target: Any):
        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('poison')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """Apply poison damage (scales heavily with stacks)"""
        # Poison scales multiplicatively with stacks
        damage = self.damage_per_second * (self.stacks ** 1.2) * dt
        if hasattr(target, 'take_damage'):
            target.take_damage(damage, 'poison')
        elif hasattr(target, 'current_health'):
            target.current_health = max(0, target.current_health - damage)


# ============================================================================
# CROWD CONTROL (CC) EFFECTS
# ============================================================================

class FreezeEffect(StatusEffect):
    """Completely immobilizes target"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="freeze",
            name="Frozen",
            duration=duration,
            duration_remaining=duration,
            max_stacks=1,  # Freeze doesn't stack
            source=source,
            params=params
        )
        self.stored_speed = 0.0

    def on_apply(self, target: Any):
        """Store original speed and set to 0"""
        if hasattr(target, 'speed'):
            self.stored_speed = target.speed
            target.speed = 0.0
        elif hasattr(target, 'movement_speed'):
            self.stored_speed = target.movement_speed
            target.movement_speed = 0.0

        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('freeze')

        # Set frozen flag for AI
        if hasattr(target, 'is_frozen'):
            target.is_frozen = True

    def on_remove(self, target: Any):
        """Restore original speed"""
        if hasattr(target, 'speed'):
            target.speed = self.stored_speed
        elif hasattr(target, 'movement_speed'):
            target.movement_speed = self.stored_speed

        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('freeze')

        if hasattr(target, 'is_frozen'):
            target.is_frozen = False

    def _apply_periodic_effect(self, dt: float, target: Any):
        """No periodic effect, just immobilization"""
        pass


class SlowEffect(StatusEffect):
    """Reduces movement speed"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="slow",
            name="Slowed",
            duration=duration,
            duration_remaining=duration,
            max_stacks=params.get('slow_max_stacks', 1),  # Usually doesn't stack
            source=source,
            params=params
        )
        self.slow_percent = params.get('slow_percent', 0.5)  # 50% slow by default
        self.stored_speed = 0.0

    def on_apply(self, target: Any):
        """Reduce speed"""
        if hasattr(target, 'speed'):
            self.stored_speed = target.speed
            target.speed *= (1.0 - self.slow_percent)
        elif hasattr(target, 'movement_speed'):
            self.stored_speed = target.movement_speed
            target.movement_speed *= (1.0 - self.slow_percent)

        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('slow')

    def on_remove(self, target: Any):
        """Restore speed"""
        if hasattr(target, 'speed'):
            target.speed = self.stored_speed
        elif hasattr(target, 'movement_speed'):
            target.movement_speed = self.stored_speed

        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('slow')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """No periodic effect"""
        pass


class StunEffect(StatusEffect):
    """Prevents all actions"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="stun",
            name="Stunned",
            duration=duration,
            duration_remaining=duration,
            max_stacks=1,  # Stun doesn't stack
            source=source,
            params=params
        )

    def on_apply(self, target: Any):
        """Set stunned flag"""
        if hasattr(target, 'is_stunned'):
            target.is_stunned = True

        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('stun')

    def on_remove(self, target: Any):
        """Remove stunned flag"""
        if hasattr(target, 'is_stunned'):
            target.is_stunned = False

        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('stun')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """No periodic effect"""
        pass


class RootEffect(StatusEffect):
    """Prevents movement but allows actions"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="root",
            name="Rooted",
            duration=duration,
            duration_remaining=duration,
            max_stacks=1,
            source=source,
            params=params
        )
        self.stored_speed = 0.0

    def on_apply(self, target: Any):
        """Set speed to 0"""
        if hasattr(target, 'speed'):
            self.stored_speed = target.speed
            target.speed = 0.0
        elif hasattr(target, 'movement_speed'):
            self.stored_speed = target.movement_speed
            target.movement_speed = 0.0

        if hasattr(target, 'is_rooted'):
            target.is_rooted = True

        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('root')

    def on_remove(self, target: Any):
        """Restore speed"""
        if hasattr(target, 'speed'):
            target.speed = self.stored_speed
        elif hasattr(target, 'movement_speed'):
            target.movement_speed = self.stored_speed

        if hasattr(target, 'is_rooted'):
            target.is_rooted = False

        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('root')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """No periodic effect"""
        pass


# ============================================================================
# BUFF EFFECTS (Positive)
# ============================================================================

class RegenerationEffect(StatusEffect):
    """Heal over time"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="regeneration",
            name="Regenerating",
            duration=duration,
            duration_remaining=duration,
            max_stacks=params.get('regen_max_stacks', 3),
            source=source,
            params=params
        )
        self.heal_per_second = params.get('regen_heal_per_second', 5.0)

    def on_apply(self, target: Any):
        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('regen')

    def on_remove(self, target: Any):
        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('regen')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """Apply healing"""
        healing = self.heal_per_second * self.stacks * dt

        if hasattr(target, 'heal'):
            target.heal(healing)
        elif hasattr(target, 'current_health') and hasattr(target, 'max_health'):
            target.current_health = min(target.current_health + healing, target.max_health)


class ShieldEffect(StatusEffect):
    """Temporary shield that absorbs damage"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="shield",
            name="Shielded",
            duration=duration,
            duration_remaining=duration,
            max_stacks=1,  # Shields don't stack, they add to shield value
            source=source,
            params=params
        )
        self.shield_amount = params.get('shield_amount', 50.0)
        self.current_shield = self.shield_amount

    def on_apply(self, target: Any):
        """Add shield to target"""
        if hasattr(target, 'shield_health'):
            target.shield_health += self.shield_amount
        else:
            # Create shield attribute if it doesn't exist
            target.shield_health = self.shield_amount

        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('shield')

    def on_remove(self, target: Any):
        """Remove shield from target"""
        if hasattr(target, 'shield_health'):
            target.shield_health = max(0, target.shield_health - self.current_shield)

        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('shield')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """No periodic effect, shield is consumed by damage"""
        # Update current shield value from target
        if hasattr(target, 'shield_health'):
            self.current_shield = min(self.shield_amount, target.shield_health)


class HasteEffect(StatusEffect):
    """Increases movement and attack speed"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="haste",
            name="Hasted",
            duration=duration,
            duration_remaining=duration,
            max_stacks=1,
            source=source,
            params=params
        )
        self.speed_bonus = params.get('haste_speed_bonus', 0.3)  # 30% faster
        self.original_speed = 0.0
        self.original_attack_speed = 0.0

    def on_apply(self, target: Any):
        """Increase speed"""
        if hasattr(target, 'speed'):
            self.original_speed = target.speed
            target.speed *= (1.0 + self.speed_bonus)
        elif hasattr(target, 'movement_speed'):
            self.original_speed = target.movement_speed
            target.movement_speed *= (1.0 + self.speed_bonus)

        if hasattr(target, 'attack_speed'):
            self.original_attack_speed = target.attack_speed
            target.attack_speed *= (1.0 + self.speed_bonus)

        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('haste')

    def on_remove(self, target: Any):
        """Restore speed"""
        if hasattr(target, 'speed'):
            target.speed = self.original_speed
        elif hasattr(target, 'movement_speed'):
            target.movement_speed = self.original_speed

        if hasattr(target, 'attack_speed'):
            target.attack_speed = self.original_attack_speed

        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('haste')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """No periodic effect"""
        pass


# ============================================================================
# DEBUFF EFFECTS (Negative stat modifiers)
# ============================================================================

class WeakenEffect(StatusEffect):
    """Reduces damage dealt"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="weaken",
            name="Weakened",
            duration=duration,
            duration_remaining=duration,
            max_stacks=params.get('weaken_max_stacks', 3),
            source=source,
            params=params
        )
        self.damage_reduction = params.get('weaken_percent', 0.25)  # 25% less damage

    def on_apply(self, target: Any):
        if not hasattr(target, 'damage_multiplier'):
            target.damage_multiplier = 1.0
        target.damage_multiplier *= (1.0 - self.damage_reduction)

        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('weaken')

    def on_remove(self, target: Any):
        if hasattr(target, 'damage_multiplier'):
            target.damage_multiplier /= (1.0 - self.damage_reduction)

        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('weaken')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """No periodic effect"""
        pass


class VulnerableEffect(StatusEffect):
    """Increases damage taken"""

    def __init__(self, duration: float, params: Dict[str, Any], source: Any = None):
        super().__init__(
            status_id="vulnerable",
            name="Vulnerable",
            duration=duration,
            duration_remaining=duration,
            max_stacks=params.get('vulnerable_max_stacks', 3),
            source=source,
            params=params
        )
        self.damage_increase = params.get('vulnerable_percent', 0.25)  # 25% more damage taken

    def on_apply(self, target: Any):
        if not hasattr(target, 'damage_taken_multiplier'):
            target.damage_taken_multiplier = 1.0
        target.damage_taken_multiplier *= (1.0 + self.damage_increase)

        if hasattr(target, 'visual_effects'):
            target.visual_effects.add('vulnerable')

    def on_remove(self, target: Any):
        if hasattr(target, 'damage_taken_multiplier'):
            target.damage_taken_multiplier /= (1.0 + self.damage_increase)

        if hasattr(target, 'visual_effects'):
            target.visual_effects.remove('vulnerable')

    def _apply_periodic_effect(self, dt: float, target: Any):
        """No periodic effect"""
        pass


# ============================================================================
# STATUS EFFECT FACTORY
# ============================================================================

STATUS_EFFECT_CLASSES = {
    'burn': BurnEffect,
    'bleed': BleedEffect,
    'poison': PoisonEffect,
    'poison_status': PoisonEffect,  # Alias
    'freeze': FreezeEffect,
    'slow': SlowEffect,
    'chill': SlowEffect,  # Alias for slow
    'stun': StunEffect,
    'root': RootEffect,
    'regeneration': RegenerationEffect,
    'regen': RegenerationEffect,  # Alias
    'shield': ShieldEffect,
    'barrier': ShieldEffect,  # Alias
    'haste': HasteEffect,
    'quicken': HasteEffect,  # Alias
    'weaken': WeakenEffect,
    'vulnerable': VulnerableEffect,
}


def create_status_effect(status_tag: str, params: Dict[str, Any], source: Any = None) -> Optional[StatusEffect]:
    """
    Factory function to create status effects from tags

    Args:
        status_tag: Tag identifier (e.g., 'burn', 'freeze')
        params: Parameters dict containing duration and effect-specific params
        source: Source entity that applied the effect

    Returns:
        StatusEffect instance or None if unknown tag
    """
    effect_class = STATUS_EFFECT_CLASSES.get(status_tag)

    if not effect_class:
        return None

    # Get duration from params (with defaults)
    duration = params.get(f'{status_tag}_duration', params.get('duration', 5.0))

    return effect_class(duration=duration, params=params, source=source)

File: Unity/Assets/Scripts/Game1.Entities/StatusEffects/StatusEffect.cs
// Game1.Entities.StatusEffects.StatusEffect
// Migrated from: entities/status_effect.py
// Phase: 3 - Entity Layer

using System;
using System.Collections.Generic;

namespace Game1.Entities.StatusEffects
{
    /// <summary>
    /// Interface for entities that can receive status effects.
    /// Both Character and Enemy implement this.
    /// </summary>
    public interface IStatusTarget
    {
        string Name { get; }
        float CurrentHealth { get; set; }
        float MaxHealth { get; }
        bool IsAlive { get; }
        float DamageMultiplier { get; set; }
        float DamageTakenMultiplier { get; set; }
        float ShieldHealth { get; set; }
        bool IsFrozen { get; set; }
        bool IsStunned { get; set; }
        bool IsRooted { get; set; }
        HashSet<string> VisualEffects { get; }
    }

    /// <summary>
    /// Base class for all status effects.
    /// Status effects have duration, stacking, tick behavior, and on_apply/on_remove hooks.
    /// </summary>
    public abstract class StatusEffect
    {
        public string StatusId { get; protected set; }
        public string DisplayName { get; protected set; }
        public float Duration { get; protected set; }
        public float TimeRemaining { get; protected set; }
        public int Stacks { get; protected set; } = 1;
        public int MaxStacks { get; protected set; } = 5;
        public float TickInterval { get; protected set; } = 1.0f;
        public Dictionary<string, object> Params { get; protected set; }
        public object Source { get; protected set; }

        private float _tickTimer;

        protected StatusEffect(string statusId, string displayName, Dictionary<string, object> parameters, object source)
        {
            StatusId = statusId;
            DisplayName = displayName;
            Params = parameters ?? new Dictionary<string, object>();
            Source = source;

            Duration = GetFloatParam("duration", 5.0f);
            TimeRemaining = Duration;
            _tickTimer = 0f;
        }

        /// <summary>
        /// Called when effect is first applied.
        /// </summary>
        public virtual void OnApply(IStatusTarget target)
        {
            target.VisualEffects.Add(StatusId);
        }

        /// <summary>
        /// Called when effect is removed (expired or dispelled).
        /// </summary>
        public virtual void OnRemove(IStatusTarget target)
        {
            target.VisualEffects.Remove(StatusId);
        }

        /// <summary>
        /// Called every tick interval. Override for DoT/HoT effects.
        /// </summary>
        protected virtual void OnTick(IStatusTarget target) { }

        /// <summary>
        /// Update the effect. Returns false if expired.
        /// </summary>
        public bool Update(float dt, IStatusTarget target)
        {
            TimeRemaining -= dt;

            // Tick-based updates
            _tickTimer += dt;
            if (_tickTimer >= TickInterval)
            {
                _tickTimer -= TickInterval;
                OnTick(target);
            }

            return TimeRemaining > 0;
        }

        /// <summary>
        /// Add stacks (up to max).
        /// </summary>
        public void AddStack(int count = 1)
        {
            Stacks = Math.Min(Stacks + count, MaxStacks);
        }

        /// <summary>
        /// Refresh duration to original value.
        /// </summary>
        public void RefreshDuration()
        {
            TimeRemaining = Duration;
        }

        // Helper methods for reading params
        protected float GetFloatParam(string key, float defaultValue = 0f)
        {
            if (Params.TryGetValue(key, out var val))
            {
                try { return Convert.ToSingle(val); }
                catch { return defaultValue; }
            }
            // Also check with status_id prefix (e.g., "burn_duration")
            string prefixedKey = $"{StatusId}_{key}";
            if (Params.TryGetValue(prefixedKey, out val))
            {
                try { return Convert.ToSingle(val); }
                catch { return defaultValue; }
            }
            return defaultValue;
        }

        protected int GetIntParam(string key, int defaultValue = 0)
        {
            if (Params.TryGetValue(key, out var val))
            {
                try { return Convert.ToInt32(val); }
                catch { return defaultValue; }
            }
            return defaultValue;
        }

        protected string GetStringParam(string key, string defaultValue = "")
        {
            if (Params.TryGetValue(key, out var val))
                return val?.ToString() ?? defaultValue;
            return defaultValue;
        }
    }

    // ========================================================================
    // DoT EFFECTS (Damage over Time)
    // ========================================================================

    public class BurnEffect : StatusEffect
    {
        private float _damagePerSecond;

        public BurnEffect(Dictionary<string, object> parameters, object source)
            : base("burn", "Burn", parameters, source)
        {
            _damagePerSecond = GetFloatParam("burn_damage_per_second",
                GetFloatParam("damage_per_second", 8.0f));
            TickInterval = 1.0f;
        }

        public override void OnApply(IStatusTarget target)
        {
            base.OnApply(target);
        }

        protected override void OnTick(IStatusTarget target)
        {
            float damage = _damagePerSecond * Stacks;
            target.CurrentHealth -= damage;
        }
    }

    public class BleedEffect : StatusEffect
    {
        private float _damagePerSecond;

        public BleedEffect(Dictionary<string, object> parameters, object source)
            : base("bleed", "Bleed", parameters, source)
        {
            _damagePerSecond = GetFloatParam("bleed_damage_per_second",
                GetFloatParam("damage_per_second", 5.0f));
            TickInterval = 1.0f;
        }

        protected override void OnTick(IStatusTarget target)
        {
            float damage = _damagePerSecond * Stacks;
            target.CurrentHealth -= damage;
        }
    }

    public class PoisonEffect : StatusEffect
    {
        private float _damagePerSecond;

        public PoisonEffect(Dictionary<string, object> parameters, object source)
            : base("poison", "Poison", parameters, source)
        {
            _damagePerSecond = GetFloatParam("poison_damage_per_second",
                GetFloatParam("damage_per_second", 4.0f));
            TickInterval = 1.0f;
        }

        protected override void OnTick(IStatusTarget target)
        {
            float damage = _damagePerSecond * Stacks;
            target.CurrentHealth -= damage;
        }
    }

    public class PoisonStatusEffect : StatusEffect
    {
        private float _damagePerSecond;

        public PoisonStatusEffect(Dictionary<string, object> parameters, object source)
            : base("poison_status", "Poison", parameters, source)
        {
            _damagePerSecond = GetFloatParam("damage_per_second", 4.0f);
            TickInterval = 1.0f;
        }

        protected override void OnTick(IStatusTarget target)
        {
            float damage = _damagePerSecond * Stacks;
            target.CurrentHealth -= damage;
        }
    }

    public class ShockEffect : StatusEffect
    {
        private float _damagePerSecond;

        public ShockEffect(Dictionary<string, object> parameters, object source)
            : base("shock", "Shock", parameters, source)
        {
            _damagePerSecond = GetFloatParam("shock_damage_per_second",
                GetFloatParam("damage_per_second", 6.0f));
            TickInterval = 1.0f;
        }

        protected override void OnTick(IStatusTarget target)
        {
            float damage = _damagePerSecond * Stacks;
            target.CurrentHealth -= damage;
        }
    }

    // ========================================================================
    // CC EFFECTS (Crowd Control)
    // ========================================================================

    public class FreezeEffect : StatusEffect
    {
        public FreezeEffect(Dictionary<string, object> parameters, object source)
            : base("freeze", "Frozen", parameters, source)
        {
            Duration = GetFloatParam("freeze_duration", GetFloatParam("duration", 3.0f));
            TimeRemaining = Duration;
        }

        public override void OnApply(IStatusTarget target)
        {
            base.OnApply(target);
            target.IsFrozen = true;
        }

        public override void OnRemove(IStatusTarget target)
        {
            base.OnRemove(target);
            target.IsFrozen = false;
        }
    }

    public class StunEffect : StatusEffect
    {
        public StunEffect(Dictionary<string, object> parameters, object source)
            : base("stun", "Stunned", parameters, source)
        {
            Duration = GetFloatParam("stun_duration", GetFloatParam("duration", 2.0f));
            TimeRemaining = Duration;
        }

        public override void OnApply(IStatusTarget target)
        {
            base.OnApply(target);
            target.IsStunned = true;
        }

        public override void OnRemove(IStatusTarget target)
        {
            base.OnRemove(target);
            target.IsStunned = false;
        }
    }

    public class RootEffect : StatusEffect
    {
        public RootEffect(Dictionary<string, object> parameters, object source)
            : base("root", "Rooted", parameters, source)
        {
            Duration = GetFloatParam("root_duration", GetFloatParam("duration", 3.0f));
            TimeRemaining = Duration;
        }

        public override void OnApply(IStatusTarget target)
        {
            base.OnApply(target);
            target.IsRooted = true;
        }

        public override void OnRemove(IStatusTarget target)
        {
            base.OnRemove(target);
            target.IsRooted = false;
        }
    }

    public class SlowEffect : StatusEffect
    {
        public SlowEffect(Dictionary<string, object> parameters, object source)
            : base("slow", "Slowed", parameters, source)
        {
            Duration = GetFloatParam("slow_duration", GetFloatParam("duration", 5.0f));
            TimeRemaining = Duration;
        }

        public override void OnApply(IStatusTarget target) => base.OnApply(target);
        public override void OnRemove(IStatusTarget target) => base.OnRemove(target);
    }

    public class ChillEffect : StatusEffect
    {
        public ChillEffect(Dictionary<string, object> parameters, object source)
            : base("chill", "Chilled", parameters, source)
        {
            Duration = GetFloatParam("chill_duration", GetFloatParam("duration", 5.0f));
            TimeRemaining = Duration;
        }

        public override void OnApply(IStatusTarget target) => base.OnApply(target);
        public override void OnRemove(IStatusTarget target) => base.OnRemove(target);
    }

    // ========================================================================
    // BUFF EFFECTS
    // ========================================================================

    public class RegenerationEffect : StatusEffect
    {
        private float _healPerSecond;

        public RegenerationEffect(Dictionary<string, object> parameters, object source)
            : base("regeneration", "Regeneration", parameters, source)
        {
            _healPerSecond = GetFloatParam("heal_per_second",
                GetFloatParam("regen_amount", 5.0f));
            TickInterval = 1.0f;
        }

        protected override void OnTick(IStatusTarget target)
        {
            float heal = _healPerSecond * Stacks;
            target.CurrentHealth = Math.Min(target.MaxHealth, target.CurrentHealth + heal);
        }
    }

    public class RegenEffect : StatusEffect
    {
        private float _healPerSecond;

        public RegenEffect(Dictionary<string, object> parameters, object source)
            : base("regen", "Regen", parameters, source)
        {
            _healPerSecond = GetFloatParam("heal_per_second", 5.0f);
            TickInterval = 1.0f;
        }

        protected override void OnTick(IStatusTarget target)
        {
            float heal = _healPerSecond * Stacks;
            target.CurrentHealth = Math.Min(target.MaxHealth, target.CurrentHealth + heal);
        }
    }

    public class ShieldEffect : StatusEffect
    {
        private float _shieldAmount;

        public ShieldEffect(Dictionary<string, object> parameters, object source)
            : base("shield", "Shield", parameters, source)
        {
            _shieldAmount = GetFloatParam("shield_amount", 50.0f);
        }

        public override void OnApply(IStatusTarget target)
        {
            base.OnApply(target);
            target.ShieldHealth += _shieldAmount * Stacks;
        }

        public override void OnRemove(IStatusTarget target)
        {
            base.OnRemove(target);
            // Don't remove shield health on expire — it was already consumed or decayed
        }
    }

    public class BarrierEffect : StatusEffect
    {
        private float _barrierAmount;

        public BarrierEffect(Dictionary<string, object> parameters, object source)
            : base("barrier", "Barrier", parameters, source)
        {
            _barrierAmount = GetFloatParam("barrier_amount", 50.0f);
        }

        public override void OnApply(IStatusTarget target)
        {
            base.OnApply(target);
            target.ShieldHealth += _barrierAmount * Stacks;
        }

        public override void OnRemove(IStatusTarget target) => base.OnRemove(target);
    }

    public class HasteEffect : StatusEffect
    {
        public HasteEffect(Dictionary<string, object> parameters, object source)
            : base("haste", "Haste", parameters, source) { }

        public override void OnApply(IStatusTarget target) => base.OnApply(target);
        public override void OnRemove(IStatusTarget target) => base.OnRemove(target);
    }

    public class QuickenEffect : StatusEffect
    {
        public QuickenEffect(Dictionary<string, object> parameters, object source)
            : base("quicken", "Quickened", parameters, source) { }

        public override void OnApply(IStatusTarget target) => base.OnApply(target);
        public override void OnRemove(IStatusTarget target) => base.OnRemove(target);
    }

    // ========================================================================
    // DEBUFF EFFECTS
    // ========================================================================

    public class WeakenEffect : StatusEffect
    {
        private float _damageReduction;

        public WeakenEffect(Dictionary<string, object> parameters, object source)
            : base("weaken", "Weakened", parameters, source)
        {
            _damageReduction = GetFloatParam("damage_reduction", 0.25f);
        }

        public override void OnApply(IStatusTarget target)
        {
            base.OnApply(target);
            target.DamageMultiplier *= (1.0f - _damageReduction * Stacks);
        }

        public override void OnRemove(IStatusTarget target)
        {
            base.OnRemove(target);
            target.DamageMultiplier /= (1.0f - _damageReduction * Stacks);
        }
    }

    public class VulnerableEffect : StatusEffect
    {
        private float _damageTakenIncrease;

        public VulnerableEffect(Dictionary<string, object> parameters, object source)
            : base("vulnerable", "Vulnerable", parameters, source)
        {
            _damageTakenIncrease = GetFloatParam("damage_taken_increase", 0.25f);
        }

        public override void OnApply(IStatusTarget target)
        {
            base.OnApply(target);
            target.DamageTakenMultiplier *= (1.0f + _damageTakenIncrease * Stacks);
        }

        public override void OnRemove(IStatusTarget target)
        {
            base.OnRemove(target);
            target.DamageTakenMultiplier /= (1.0f + _damageTakenIncrease * Stacks);
        }
    }

    // ========================================================================
    // FACTORY
    // ========================================================================

    /// <summary>
    /// Factory for creating status effects from tag strings.
    /// Matches Python create_status_effect() function.
    /// </summary>
    public static class StatusEffectFactory
    {
        public static StatusEffect Create(string statusTag, Dictionary<string, object> parameters, object source = null)
        {
            return statusTag.ToLower() switch
            {
                "burn"          => new BurnEffect(parameters, source),
                "bleed"         => new BleedEffect(parameters, source),
                "poison"        => new PoisonEffect(parameters, source),
                "poison_status" => new PoisonStatusEffect(parameters, source),
                "shock"         => new ShockEffect(parameters, source),
                "freeze"        => new FreezeEffect(parameters, source),
                "stun"          => new StunEffect(parameters, source),
                "root"          => new RootEffect(parameters, source),
                "slow"          => new SlowEffect(parameters, source),
                "chill"         => new ChillEffect(parameters, source),
                "regeneration"  => new RegenerationEffect(parameters, source),
                "regen"         => new RegenEffect(parameters, source),
                "shield"        => new ShieldEffect(parameters, source),
                "barrier"       => new BarrierEffect(parameters, source),
                "haste"         => new HasteEffect(parameters, source),
                "quicken"       => new QuickenEffect(parameters, source),
                "weaken"        => new WeakenEffect(parameters, source),
                "vulnerable"    => new VulnerableEffect(parameters, source),
                _               => null
            };
        }
    }
}
